### PR TITLE
LocalPlayer: Restore 2u height sneak jump

### LINKED ISSUE #15008: With both new_move and sneak_glitch enabled you can no longer sneak+jump 2 blocks
### Minetest version

Any after df8a600b228c93e0a2ab1ef40dfa250b0c994718

### Summary

Expected behavior:

https://github.com/user-attachments/assets/8b2e635b-89e5-43c2-b90e-ae43597c96ee

It simply doesn't work now

### Steps to reproduce

Run
```
minetest.get_player_by_name("singleplayer"):set_physics_override({sneak_glitch = true, new_move = true})
```
in a world via some sort of mod/lua command

Try to do what's shown in the video

### PR DESCRIPTION
**Pull request by SmallJoker**
Fix 1: Do not consider LocalPlayer's CAO in the collision data.
Fix 2: work around the "aabbox3d::intersectsWithBox" edge-case.

Fixes #15008, caused by #14332

## To do

This PR is Ready for Review.

## How to test

1. https://content.minetest.net/packages/Krock/sneak_glitch/
2. Build a 2 node pillar
3. Hold jump, then hold sneak
4. Node pillar must be climbable
5. Ensure that the ceiling is still detected when using the sneak ladder. See https://github.com/minetest/minetest/pull/14332#issue-2110951340 for a demo.

### CODE REVIEW

File: src/collision.cpp [hunk 1]
```diff
@@ -623,8 +623,10 @@ bool collision_check_intersection(Environment *env, IGameDef *gamedef,
 		Collision detection
 	*/
 	aabb3f checkbox = box_0;
-	checkbox.MinEdge += pos_f;
-	checkbox.MaxEdge += pos_f;
+	// aabbox3d::intersectsWithBox(box) returns true when the faces are touching perfectly.
+	// However, we do not want want a true-ish return value in that case. Add some tolerance.
```
[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/collision.cpp:627]: Alternatively I could implement `intersectsWithBoxButFacesExclusive(box)` where checks are performed with `>` instead of `>=`. Conveniently, the tolerance also compensates for any floating point imprecisions that *might* still occur (needs testing).

[the remark this answers is not in the document]
Comment (grorp) [reply src/collision.cpp:627]: > this change sounds potentially far-reaching

`collision_check_intersection` is only used by the sneak glitch "fix glitch through ceiling" thing though.

This behavior needs documentation in the header file.

[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/collision.cpp:627]: @grorp Documented in 04fd8aea8.

File: src/collision.cpp [hunk 2]
```diff
@@ -301,12 +301,12 @@ static void add_object_boxes(Environment *env,
 
 		// add collision with local player
 		LocalPlayer *lplayer = c_env->getLocalPlayer();
-		if (lplayer->getParent() == nullptr) {
+		auto *obj = (ActiveObject*) lplayer->getCAO();
+		if (self != obj && !lplayer->getParent()) {
```
[the remark this answers is not in the document]
Comment (grorp) [reply src/collision.cpp:305]: Confirmed that LocalPlayer collides with itself, using this patch that makes walking on air possible:

```diff
diff --git a/src/collision.cpp b/src/collision.cpp
index d804f7f0d..01738cadb 100644
--- a/src/collision.cpp
+++ b/src/collision.cpp
@@ -306,6 +306,8 @@ static void add_object_boxes(Environment *env,
                        v3f lplayer_pos = lplayer->getPosition();
                        lplayer_collisionbox.MinEdge += lplayer_pos;
                        lplayer_collisionbox.MaxEdge += lplayer_pos;
+                       lplayer_collisionbox.MaxEdge.Y = lplayer_collisionbox.MinEdge.Y;
+                       lplayer_collisionbox.MinEdge.Y -= 0.5f;
                        auto *obj = (ActiveObject*) lplayer->getCAO();
                        cinfo.emplace_back(obj, 0, lplayer_collisionbox);
                }
```

[the remark this answers is not in the document]
Comment (grorp) [reply src/collision.cpp:305]: I propose these changes: https://github.com/minetest/minetest/commit/1a442e6da67c7ef62ef1d615db9af295ad856541. This fixes the self-collision issue in both cases.

[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/collision.cpp:305]: Currently it seems to me that the parameter `self` specifies the ActiveObject that must be ignored to avoid colliding with itself. This PR extends this idea for LocalPlayer's CAO.

Counter-proposal:
```diff
diff --git a/src/client/localplayer.cpp b/src/client/localplayer.cpp
index 51e91ab94..75f755578 100644
--- a/src/client/localplayer.cpp
+++ b/src/client/localplayer.cpp
@@ -360,7 +360,7 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
 
        collisionMoveResult result = collisionMoveSimple(env, m_client,
                pos_max_d, m_collisionbox, player_stepheight, dtime,
-               &position, &m_speed, accel_f);
+               &position, &m_speed, accel_f, m_cao);
 
        bool could_sneak = control.sneak && !free_move && !in_liquid &&
                !is_climbing && physics_override.sneak;
@@ -990,7 +990,7 @@ void LocalPlayer::old_move(f32 dtime, Environment *env, f32 pos_max_d,
 
        collisionMoveResult result = collisionMoveSimple(env, m_client,
                pos_max_d, m_collisionbox, player_stepheight, dtime,
-               &position, &m_speed, accel_f);
+               &position, &m_speed, accel_f, m_cao);
 
        // Position was slightly changed; update standing node pos
        if (touching_ground)
@@ -1254,7 +1254,7 @@ void LocalPlayer::handleAutojump(f32 dtime, Environment *env,
 
        // try at peak of jump, zero step height
        collisionMoveResult jump_result = collisionMoveSimple(env, m_client, pos_max_d,
-               m_collisionbox, 0.0f, dtime, &jump_pos, &jump_speed, v3f(0.0f));
+               m_collisionbox, 0.0f, dtime, &jump_pos, &jump_speed, v3f(0.0f), m_cao);
 
        // see if we can get a little bit farther horizontally if we had
        // jumped
```

File: src/collision.h [hunk 3]
```diff
@@ -73,7 +73,9 @@ collisionMoveResult collisionMoveSimple(Environment *env,IGameDef *gamedef,
 		v3f accel_f, ActiveObject *self=NULL,
 		bool collide_with_objects=true);
 
-// check if box is in collision on actual position
+/// @param self (optional) ActiveObject to ignore in the calculation.
+/// @returns `true` when `box_0` truely intersects with a node or object.
```
Comment (grorp) [inline src/collision.h:77]:
```suggestion
/// @returns `true` when `box_0` truly intersects with a node or object.
```

File: src/collision.h [hunk 4]
```diff
@@ -73,7 +73,9 @@ collisionMoveResult collisionMoveSimple(Environment *env,IGameDef *gamedef,
 		v3f accel_f, ActiveObject *self=NULL,
 		bool collide_with_objects=true);
 
-// check if box is in collision on actual position
+/// @param self (optional) ActiveObject to ignore in the calculation.
```
Comment (grorp) [inline src/collision.h:76]: `collisionMoveSimple` could get this comment too

File: src/collision.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (grorp) [reply src/collision.cpp:305]: 👍 for your proposal

In two other places, the check is `self != obj && self != obj->getParent()`. Would the second part also make sense here?

[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/collision.cpp:305]: You're right. `self` must not collide with itself and not with its children.

File: src/collision.cpp [hunk 5]
```diff
@@ -301,8 +301,8 @@ static void add_object_boxes(Environment *env,
 
 		// add collision with local player
 		LocalPlayer *lplayer = c_env->getLocalPlayer();
-		auto *obj = (ActiveObject*) lplayer->getCAO();
-		if (self != obj && !lplayer->getParent()) {
+		auto *obj = (ClientActiveObject*) lplayer->getCAO();
+		if (self != obj && self != obj->getParent()) {
```
[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/collision.cpp:305]: You're right. This condition should be `(!self || (self != obj && self != obj->getParent())` like in the other conditions.

About the second question:

An example: `self == nullptr` is used by particles. They don't have an ActiveObject but can be expected to collide with objects. It should not make a difference whether LocalPlayer's CAO is attached or not.

Note: As of master, objects may collide with such that are attached, but attached objects do not perform collision checks (-> `getParent()` guards).

**Review by grorp — APPROVED** [review]
- #15008 is fixed
- Theoretical LocalPlayer self-collision issue introduced by #14582 is fixed
- Code makes sense
